### DISCUSSION TITLE
Segment ends should contain nodeId

### DISCUSSION THREAD

**Discussion by originalfoo** [CitiesSkylinesMods/TMPE#1504] [Ideas]
We've got loads of excess lines of code due to determining `StartNode` practically every time we interact with segment ends.

That's almost always determined by `segmentId.ToSegment().IsStartnode(nodeId)` - so this line is appearing over and over and over again.

If we stored `nodeId` in segment ends, and had constructor that takes `segmentId, nodeId` and works out the `StartNode` from that, we could trim lots of code bloat (at the expense of adding a `ushort` to a struct). This would remove huge numbers of `ToSegment()` lookups and `IsStartnode()` checks throughtout the codebase.

What do you think?

**Comment by krzychu124** [comment]
Not bad idea if we could ensure or assume that id or rather its connection with segment won't change (always be valid). I think we should also try to merge obsolete `SegmentEnd`(Traffic.Impl) with `ExtSegment` but they have slightly different id/index generation :/

**Comment by originalfoo** [comment]
I added the constructor that takes nodeId in a recent PR #1506; it converts to `startNode` and stores that - just makes code a little cleaner:

https://github.com/CitiesSkylinesMods/TMPE/pull/1506/files#diff-977b7061ebbe786496caff8a3e924dd5e61a8ca30d063a4b12e897e073ab8f8a

**Reply by krzychu124** [reply]
oh, right, missed that change🤔
